### ISSUE TITLE
Question on how you initialize and update states  ?

### ISSUE THREAD

**Issue by ghost** [jbornschein/draw#8]
Hi Jorg,

I was just wondering how you implement _state initialization and update_, in your code, and how this is done in Blocks in general, (I'm still a novice with both Theano and Blocks). I think this is related to how you get different digits, every time you sample with the trained decoder?

To be more precise, when you start the main loop, for its first iteration on the first batch, in the first epoch, what values do you use for, 

`c`, `h_enc`, `c_enc`, `h_dec`, `c_dec` 

I've tried a couple of things - 

1) just set `h_enc`, `c_enc`, `h_dec`, `c_dec` to zero, at the start of each sequence of iterations of an SGD step. I'm guessing this is what your code does, from the Blocks [RNN docs](https://github.com/bartvm/blocks/blob/master/docs/rnn.rst). I set `c` initially to -10, thus it's sigmoid is zero, i.e. a blank canvas.  

2) so called backpropogation through time, (BPTT), where at the end of one sequence of iterations, `t = 1,...,T`, I set 

`h_enc_0 = h_enc_T` , `h_dec_0 = h_dec_T` , etc

That is, I use the final states of the encoder and decoder's hidden outputs and cell values, from the previous SGD step, as the initial states of the encoder and decoder, hidden outputs and cell values, for the first iteration of the objective function used in the present SGD step. I also tried to do the same for the canvas matrix 'c', but that doesn't really make sense - i.e. why write on top of an old "canvas" which already has an image written on it?
- Is some similar sort of state transfer of `h_enc`, `c_enc`, `h_dec`, `c_dec` implemented in the Blocks, LSTM module, and your code here? That is, are the initial states for the first iteration of the current SGD step, set to the final states calculated by the last iteration of the previous SGD step?

Basically I'm still having problems generating convincing looking digits from my system with all of these variants - it confuses me which is the 'right' one theoretically - I'm not sure I understand how BPTT, can be applied to the DRAW system of networks? 
- A related question is - which method do you use when you sample the batch of 16*16 canvases with the decoder, with `sample.py`? 

I can't see where in your code, or how Theano/Blocks does this? Basically your first sampled/reconstructed canvas, i.e. `samples-000.png` seems to already contain some biases which then "evolve", to form the final images. I seem to have missed something in my implementation as my initial canvases, `c_0` are simply blank.

I think this relates to a line in the [paper](http://arxiv.org/pdf/1502.04623.pdf), (just above eqn 3)

>  For each image x presented to the network, `c_0` , `h^enc_0` , `h^dec_0` , are initialised to learned biases.

May I ask how you implement this in your code? It's been confusing me for weeks? I've tried adding a simple bias layer that takes, `c_0` , `h^enc_0` , `h^dec_0` as inputs, but it didn't make much difference?

Sorry for the long question, but I'm trying to understand your code, with a broken 'Blocks` installation :(

Best Regards,

Aj  

PS - my experiments are still all with the without attention version of the DRAW paper, for small numbers of iterations (~4), and also sometimes using the same sample from the prior for all iterations..

**Comment by ghost** [comment]
Sorry Jorg, 

this was a bit of a dumb question. It turns out I made a few mistakes in my code and it was driving me nuts - I started grasping at straws trying to fix it. Here's what I did wrong - 
- In eqn 5, I forgot to concatenate the read vector `r_t` with the previous decoder output `h_dec_t-1`
- I'm using larger batch sizes, ~128 and ~256 to control the variance of the BCE loss
- I'm using batch norm on the encoders output to control the variance of the KLD loss
- I was using a fudge for the gradient of the KLD loss, (you need to do this in Torch), which I've now fixed

Along with working through your code and learning more about `Blocks`, (if I had to start again I'd use it rather than Torch), together all of these things seem to have fixed and improved my code. 

I think `Blocks` is really cool, it's beautifully designed, and I definitely will use it more in the future!

Best regards,

Aj
